### PR TITLE
User Location Banner

### PR DESCRIPTION
**Pull request by zfox23**
Allows a user to rez a "banner" entity which displays their location.

https://hifi-content.s3.amazonaws.com/Experiences/Releases/usefulUtilities/userLocation/2019-04-02_09-52-00/userLocation.json
^ Edit the `userData`.

I have to add the "request" module to this folder; the entity script can't load the request module from disk.

### CODE REVIEW

**Comment by salinacsoto** [comment]
bitmoji

File: usefulUtilities/userLocation/userLocation.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+//  Copyright 2019 High Fidelity, Inc.
+//
+//  See accompanying README.md for usage instructions.
+// 
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+
+
+(function () {
+    var request = Script.require("./modules/request.js").request;
```
Comment (MiladNazeri) [inline usefulUtilities/userLocation/userLocation.js]: Is there a reason why we can't access request in client/server?  I think we are adding a lot of the same module over and over again.  Lol do a search for request in the hifi-content code base.

File: usefulUtilities/userLocation/userLocation.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    var USER_API_BASE_URL = Account.metaverseServerURL + "/api/v1/users?filter=connections&search=";
+    var TEXT_LOCAL_ENTITY_PROPS = {
+        "type": "Text",
+        // eslint-disable-next-line no-magic-numbers
+        "localPosition": [0, 0, 0.05],
+        // eslint-disable-next-line no-magic-numbers
+        "localRotation": Quat.fromVec3Degrees([0, 0, 0]),
+        "lineHeight": 0.08,
+        // eslint-disable-next-line no-magic-numbers
+        "dimensions": [1.5, 0.2, 0.05],
```
Comment (MiladNazeri) [inline usefulUtilities/userLocation/userLocation.js]: Ok I feel better about using arrays for vecs after seeing this, but taking away that it just depends on the context and if Milad can read it.

File: usefulUtilities/userLocation/userLocation.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        "script": Script.resolvePath("userLocationTextLocalEntity.js?6")
+    };
+
+    var UserLocationBanner = function() {
+        _this = this;
+    };
+
+    UserLocationBanner.prototype = {
+        preload: function(entityID) {
+            _this.entityID = entityID;
```
Comment (MiladNazeri) [inline usefulUtilities/userLocation/userLocation.js]: I like your use of consts at the top, but properties in the constructor itself.   Stealing.

File: usefulUtilities/userLocation/userLocation.js [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            _this.targetUsername = false;
+            _this.targetDisplayName = false;
+            _this.refreshLocationInterval = false;
+            _this.textLocalEntity = false;
+            _this.canTeleportToTargetUser = false;
+    
+            var properties = Entities.getEntityProperties(entityID, ["userData"]);
```
Comment (MiladNazeri) [inline usefulUtilities/userLocation/userLocation.js]: If the only property you are getting is userdata, then maybe just save yourself a var

File: usefulUtilities/userLocation/userLocation.js [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                }
+    
+                if (userData.targetDisplayName && userData.targetDisplayName.length > 0) {
+                    _this.targetDisplayName = userData.targetDisplayName;
+                } else {
+                    console.log("Please specify `targetDisplayName` inside this entity's `userData`!");
+                    return;
+                }
+            } else {
+                console.log("Please specify this entity's `userData`! See README.md for instructions.");
```
Comment (MiladNazeri) [inline usefulUtilities/userLocation/userLocation.js]: I wish we could get feedback like this not in the console.  Maybe we can use notifications for these, or something else that doesn't depend on the console to see why there was an error.

File: usefulUtilities/userLocation/userLocation.js (hunk 1 above)
Comment (zfox23) [reply usefulUtilities/userLocation/userLocation.js]: discussed in Slack!

File: usefulUtilities/userLocation/userLocation.js (hunk 2 above)
Comment (zfox23) [reply usefulUtilities/userLocation/userLocation.js]: 👍

File: usefulUtilities/userLocation/userLocation.js (hunk 3 above)
Comment (zfox23) [reply usefulUtilities/userLocation/userLocation.js]: 👍

File: usefulUtilities/userLocation/userLocation.js (hunk 4 above)
Comment (zfox23) [reply usefulUtilities/userLocation/userLocation.js]: this feels slightly cleaner to me personally!

Comment (zfox23) [reply usefulUtilities/userLocation/userLocation.js]: but good point :)

File: usefulUtilities/userLocation/userLocation.js (hunk 5 above)
Comment (zfox23) [reply usefulUtilities/userLocation/userLocation.js]: i agree...notifications seem like the right move, if our notification system were better. i'll think about this for future projects. great idea

**Pull request merged**